File: backend/services/agent-runner/src/stigmer_runner/worker/otel.py
"""Env-var-driven OpenTelemetry tracing initialization.

Mirrors the workflow-runner pattern (``pkg/otel/otel.go``): reads
``OTEL_EXPORTER_OTLP_ENDPOINT`` and configures an OTLP/gRPC exporter
with W3C TraceContext + Baggage propagation.  When the env var is unset,
``init_tracing`` returns ``None`` — zero overhead, no SDK initialization.
"""

import logging
import os
from collections.abc import Callable

logger = logging.getLogger(__name__)


def init_tracing(service_name: str) -> Callable[[], None] | None:
    """Initialize OTel tracing if ``OTEL_EXPORTER_OTLP_ENDPOINT`` is set.

    Returns a synchronous shutdown callable that flushes buffered spans,
    or ``None`` when tracing is disabled.
    """
    endpoint = os.environ.get("OTEL_EXPORTER_OTLP_ENDPOINT")
    if not endpoint:
        return None

    from opentelemetry import trace
    from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
    from opentelemetry.sdk.resources import Resource
    from opentelemetry.sdk.trace import TracerProvider
    from opentelemetry.sdk.trace.export import BatchSpanProcessor

    resource = Resource.create({"service.name": service_name})
    exporter = OTLPSpanExporter(endpoint=endpoint, insecure=True)
    provider = TracerProvider(resource=resource)
    provider.add_span_processor(BatchSpanProcessor(exporter))
    trace.set_tracer_provider(provider)

    logger.info("OTel tracing enabled (endpoint=%s, service=%s)", endpoint, service_name)

    def shutdown() -> None:
        provider.shutdown()

    return shutdown

File: test/integration/harness/otel.go
package harness

import (
	"context"
	"fmt"

	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

const tracerName = "stigmer.integration-test"

// Tracer returns the OTel tracer for integration test spans.
// When tracing is disabled (no Jaeger), this returns a no-op tracer.
func Tracer() trace.Tracer {
	return otel.Tracer(tracerName)
}

// InitTracing configures the global OTel TracerProvider to export spans to
// a Jaeger instance at the given gRPC address (host:port). Returns a
// shutdown function that flushes remaining spans.
func InitTracing(ctx context.Context, jaegerGRPCAddress string) (shutdown func(context.Context) error, err error) {
	conn, err := grpc.NewClient(jaegerGRPCAddress,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	if err != nil {
		return nil, fmt.Errorf("dial jaeger otlp: %w", err)
	}

	exporter, err := otlptracegrpc.New(ctx, otlptracegrpc.WithGRPCConn(conn))
	if err != nil {
		return nil, fmt.Errorf("create otlp exporter: %w", err)
	}

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceNameKey.String("integration-test"),
		),
	)
	if err != nil {
		return nil, fmt.Errorf("create otel resource: %w", err)
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
	)

	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
	))

	return tp.Shutdown, nil
}

// OTelGRPCDialOptions returns gRPC dial options that add OTel client
// interceptors for automatic span creation and W3C traceparent propagation.
func OTelGRPCDialOptions() []grpc.DialOption {
	return []grpc.DialOption{
		grpc.WithStatsHandler(otelgrpc.NewClientHandler()),
	}
}
